src/bin/server/transcoder/actor.rs

```diff
@@ -0,0 +1,215 @@
+
+        let transcode_future = self
+            .transcode(&msg)
+            .map(move |(_output, temporary_file_path)| {
+                // TODO: Try Using Failure Errors
+                // Result ignored here because if inserting a file fails later,
+                // getting it out of the cache will fail also. It's better to
+                // start the error here, but forwarding errors from here is
+                // difficult because Shared<Future<...>> turns the error into
+                // an error reference. error-chain errors can't be cloned.
+                let _ = cache
+                    .borrow_mut()
+                    .insert_file(&future_map_key, temporary_file_path);
+
+                future_cache.borrow_mut().remove(&future_map_key);
+
+                ()
+            });
+
+        let boxed_transcode_future: Box<Future<Item = (), Error = io::Error>> =
+            Box::new(transcode_future);
+
+        let shared_future = boxed_transcode_future.shared();
+
+        self.future_cache
+            .borrow_mut()
+            .insert(key.clone(), shared_future.clone());
+
+        Box::new(shared_future.map(|_| ()).map_err(|err| err.kind().into()))
+    }
+
+    fn disk_cache_has(&self, key: &OsStr) -> bool {
+        self.disk_cache.borrow().contains_key(key)
+    }
+
+    fn future_cache_has(&self, key: &OsStr) -> bool {
+        self.future_cache.borrow().contains_key(key)
+    }
+
+    /// Returns a future which resolves when the transcode job associated with the message is
+    /// complete and the transcoded file is in the disk_cache.
+    fn get_cache_populated_future<P: AsRef<Path>>(
+        &self,
+        msg: &TranscodeMessage<P>,
+    ) -> ResponseFuture<(), io::Error> {
+        let key = msg.to_key();
+
+        if self.disk_cache_has(&key) {
+            return Box::new(future::ok(()));
+        }
+
+        if self.future_cache_has(&key) {
+            return Box::new(
+                self.future_cache
+                    .borrow()
+                    .get(&key)
+                    .unwrap()
+                    .clone()
+                    .map(|_| ())
+                    .map_err(|err| err.kind().into()),
+            );
+        }
+
+        Box::new(self.transcode_and_cache(&msg))
+    }
+}
+
+impl Actor for Transcoder {
+    type Context = Context<Self>;
+}
+
+impl<P> Handler<TranscodeMessage<P>> for Transcoder
+where
+    P: AsRef<Path>,
+{
+    type Result = ResponseFuture<Box<ReadSeek>, errors::Error>;
+
+    fn handle(&mut self, msg: TranscodeMessage<P>, _ctx: &mut Self::Context) -> Self::Result {
+        let disk_cache_ref = self.disk_cache.clone();
+        let key = msg.to_key();
+
+        Box::new(
+            self.get_cache_populated_future(&msg)
+                .map_err(|e| -> errors::Error { e.into() })
+                .and_then(move |_| {
+                    let mut disk_cache = disk_cache_ref.borrow_mut();
+                    let file = disk_cache
```

Comment:
There's no reason to make an extra variable here, just chain it:
```rust
let file = disk_cache_ref.borrow_mut()
    .get(&key)
    // ...
```

Reply:
There was an issue with lifetimes when done this way. `disk_cache_ref` does not live long enough.

Reply:
Hm, me too. I must have changed something.